lib/einhorn/config_file.rb

```diff
@@ -0,0 +1,182 @@
+        IO.new(fd).close
+        return
+      end
+
+      # Make sure that the desired FD is available
+      desired_config[:fd] ||= config[:fd]
+      desired_fd = desired_config[:fd]
+      desired_flags = desired_config[:flags]
+
+      # Check on whether anyone's squatting on our FD
+      holder_addr, holder_conf = current.detect {|k, conf| conf[:fd] == desired_fd}
+      if holder_addr && holder_addr == [addr, port]
+        # No binding needed here
+      elsif holder_addr
+        # Someone else is occupying this FD. Need to dup their FD and
+        # clear up the current slot.
+        holder_io = IO.new(desired_fd)
+
+        fd2 = holder_io.fcntl(Fcntl::F_DUPFD)
+        Einhorn.log_info("Clearing room for #{addr}:#{port} by renumbering fd for #{holder_addr.join(':')} from #{desired_fd} -> #{fd2}")
+
+        # Force it closed since Ruby 1.8 would do an autoclose
+        # anyway, and we can close it out for now.
+        holder_io.close
+        holder_conf[:fd] = fd2
+      end
+
+      if !config
+        # Open up our own fresh file descriptor, since the desired one
+        # must now be free.
+        actual_fd = Einhorn::FD.bind(addr, port, desired_flags)
+        # This is the only case where we might change the value of
+        # desired_config[:fd], so we can't use desired_fd here.
+        desired_config[:fd] ||= actual_fd
+        if actual_fd != desired_config[:fd]
+          Einhorn::FD.dup2flags(actual_fd, desired_config[:fd])
+          IO.new(actual_fd).close
+        end
+        return
+      end
+
+      # Check on whether we need to change our existing FD number
+      if config && (current_fd = config[:fd]) != desired_fd
+        Einhorn.log_info("Renumbering #{addr}:#{port} from #{current_fd} -> #{desired_fd}")
+
+        # Actually renumber to the current value if needed
+        current_io = IO.new(current_fd)
+        Einhorn::FD.dup2flags(current_io, desired_fd)
+        current_io.close
+        config[:fd] = desired_fd
+      end
+
+      # Check on whether we need to change our existing flags
+      if config && config[:flags] != desired_flags
+        Einhorn.log_info("Changing flags on #{addr}:#{port} (#{desired_fd}) from #{config[:flags]} to #{desired_flags}")
+
+        Einhorn::FD.set_flags(desired_fd, desired_flags)
+      end
+    end
+
+    def self.load_config!
+      return unless (file = Einhorn::State.config_file)
+      load!(file)
+    end
+
+    def self.load!(filepath)
+      es = []
+      begin
+        es << Psych::BadAlias
+      rescue NameError
+      end
+
+      begin
+        loaded = YAML.load_file(filepath)
+      rescue *es => e
+        # YAML parse-time errors include the filepath already, but
+        # load-time errors do not.
+        #
+        # Specifically, `Psych::BadAlias` (raised by doing something
+        # like `YAML.load('foo: *bar')`) does not:
+        # https://github.com/tenderlove/psych/issues/192
+        e.message << " (while loading #{filepath})"
```

Comment:
As a note, this code was borrowed from Chalk::Config. There was some discussion with @nelhage at the time about the best approach, and we both felt that sketching on error messages is unfortunate but there's not a particularly great pattern for chained exceptions in Ruby.

In general, I've yet to see a really satisfactory answer for exception wrapping in Ruby, since you need to figure out what to do about:

- The exception type. (If you create a new exception of the same type, what if it has a complex constructor? If you use a general-purpose wrapper class, now you're probably not going to log the original type.)
- The backtrace. (If you copy over the existing backtrace, then you're lying to your users about where this exception originated; if you use the backtrace where the exception was generated then you don't know where the originator came from.)

One plausible default might be "use a generic wrapper class, stick the original exception type into the message, and copy over the backtrace". Curious if you have a better one?